packages/SwingSet/src/kernel/cleanup.js

```diff
@@ -18,45 +18,94 @@ export function deleteCListEntryIfEasy(
                 if (scanKernelPromise(slot, data)) {
-                  kdebug(`@@@ scan ${slot} detects circularity`);
+                  // kdebug(`@@@ scan ${slot} detects circularity`);
                   return true;
                 }
               } else {
-                kdebug(`@@@ scan ${slot} state = ${state}`);
+                // kdebug(`@@@ scan ${slot} state = ${state}`);
               }
             }
           }
         }
       }
-      kdebug(`@@@ scan ${scanKPID} detects no circularity`);
+      // kdebug(`@@@ scan ${scanKPID} detects no circularity`);
       return false;
     }
 
-    kdebug(`@@ checking ${vatID} ${kpid} for circularity`);
+    // kdebug(`@@ checking ${vatID} ${kpid} for circularity`);
     if (scanKernelPromise(kpid, kernelData)) {
-      kdebug(
-        `Unable to delete ${vatID} clist entry ${kpid}<=>${vpid} because it is indirectly self-referential`,
-      );
+      // kdebug(
+      //  `Unable to delete ${vatID} clist entry ${kpid}<=>${vpid} because it is indirectly self-referential`,
+      // );
       return;
     } else if (sawPromise) {
-      kdebug(
-        `Unable to delete ${vatID} clist entry ${kpid}<=>${vpid} because there was a contained promise ${sawPromise}`,
-      );
+      // kdebug(
+      //  `Unable to delete ${vatID} clist entry ${kpid}<=>${vpid} because there was a contained promise ${sawPromise}`,
+      // );
       return;
     }
   }
   vatKeeper.deleteCListEntry(kpid, vpid);
 }
+
+export function getKpidsToRetire(
+  vatID,
+  vatKeeper,
+  kernelKeeper,
+  rootKPID,
+  rootKernelData,
+) {
+  const seen = new Set();
+  function scanKernelPromise(kpid, kernelData) {
+    // kdebug(`### scanning ${kpid} ${JSON.stringify(kernelData)}`);
+    if (vatKeeper.hasCListEntry(kpid)) {
+      // kdebug(`## adding ${kpid} to scan results`);
+      seen.add(kpid);
+      if (kernelData) {
+        for (const slot of kernelData.slots) {
+          const { type } = parseKernelSlot(slot);
+          // kdebug(`## examine ${kpid} slot ${slot}`);
+          if (type === 'promise') {
+            if (!seen.has(slot)) {
+              const kp = kernelKeeper.getKernelPromise(slot);
+              const { data, state } = kp;
+              // kdebug(`## state of ${slot} is: ${JSON.stringify(kp)}`);
+              if (state !== 'unresolved') {
+                if (data) {
+                  scanKernelPromise(slot, data);
+                }
+              } else {
+                // kdebug(`## ${slot} is still unresolved`);
+              }
+            } else {
+              // kdebug(`## ${slot} previously seen`);
+            }
+          } else {
+            // kdebug(`## ${slot} is not a promise`);
+          }
+        }
+      } else {
+        // kdebug(`## ${kpid} has no data`);
+      }
+    } else {
+      // kdebug(`## ${kpid} has no c-list entry for ${vatID}`);
+    }
+  }
+
+  scanKernelPromise(rootKPID, rootKernelData);
+  return Array.from(seen);
```

Comment:
`Array.from(set)` is actually `Array.from(iterable)` and relies on Set's iteration order for the array order.  The iteration order of a Set (as well as the various other collection classes and collection-like things) is specified in ECMA-262 to be insertion order, which is the order we want here because it corresponds to graph traversal order, which also means dependency order, which I think is necessary for correct functioning.  If you were to sort the array as a separate step, you'd need a comparator that could compare dependency order, which I don't see how to implement short of keeping an explicit record of the depth-first traversal order, which Set already does (you could, for example, accumulate the array as you go instead of doing an explicit sort at the end, but the scanning algorithm needs to do random-access lookups, which would require maintaining the Set also for queries, but if you do that you don't need the array due to the aforementioned Set iteration semantics).

I'd note that while depth-first graph traversal order yields a dependency ordering, depth-first order is unique while dependency order is not necessarily unique since dependency order is a lattice, so if you were to implement this in another language that didn't have something like Set with JavaScript's ordering rules, you might have to do the work yourself -- so we might need to explicitly specify that the intervat representation of a notification presents the promises in some defined dependency ordering like depth-first, even though that's just what the JavaScript path of least resistance delivered us for free.

My intuition now leans towards thinking the `resolutions` object should be changed to an array.  (I actually had the intuition it should be an array at the beginning but I let you talk me out of it because my sense that an array was better was due to a sort of cargo-cult notion that all the other things we send are arrays so this one should be too; I wasn't really thinking about the importance of ordering -- and random-access is convenient.)  As long was we are in the JavaScript universe, we are OK because JavaScript objects also obey the insertion-order-is-iteration-order rule, but once we encode things to JSON it's possible that a non-JavaScript JSON implementation might re-order things, since JSON objects are explicitly unordered (or, more pedantically, the JSON spec takes no position on whether or not a JSON implementation will impute meaning to the ordering of the properties in a JSON object) whereas JSON arrays are explicitly ordered.

Reply:
Aye. I guess it's fine to have the order determined by the graph being traversed, that's plenty deterministic.

I suppose the one benefit from using an Object or Map or something else key-value-ish is the enforced property that each promise-id can only be included once, whereas with an Array a misbehaving function might produce one with duplicate keys. But I think the fact that ordering matters here should push us into preferring the Array despite that minor drawback.. using an Array makes it abundantly clear to the reader that the creator of `resolutions` has a specific order in mind, whereas relying upon JS's (well-specified but non-obvious) insertion ordering is not nearly as discoverable. Some auditor down the road will save time if we use the Array for this, and that's enough reason to do it.